### ISSUE TITLE
Confusion regarding Voice Keyer

### ISSUE THREAD

**Issue by barjac** [drowe67/freedv-gui#438]
From what I can read/guess, the voice keyer .wav files should be unprocessed audio from the mic input.

I can see no option in the tools menu to record such a file un-digitized.

A CQ type message needs to be recorded using the same mic and signal path and levels as when transmitting normally, so it sounds seamless, but how?

I see the option to 'Start play file - mic in' but none for 'Start record file - mic in' .

What am I missing?

**Comment by Tyrbiter** [comment]
This is from a FreeDV manual at v1.1.0, which is pretty old:

"You will need to record the audio you wish to send as a WAV file using whatever program you have available on your PC to do this. Note: the audio MUST be recorded with an 8K sample rate or it will not work correctly with FreeDV."

So the long and short is that you need some other software to record the audio. It would certainly be nice if the freedv program had such a facility added to make things more straightforward.

**Comment by Tyrbiter** [comment]
How difficult would be a record from selected microphone to WAV file be to implement @tmiw ?

I would guess that the nuts and bolts to do this are already present in freedv.

**Comment by barjac** [comment]
I suspected that it was not included so thanks for confirming.
Thinking ahead...
It would be nice to be able to click to start recording during a transmission and then hit the voice keyer button to continue using the recording with the existing timings.
Quick, easy to use on the fly and levels and acoustics can match perfectly.
Files could be sequentially named and manually renamed later for re-use (somehow) should the need arise.

**Comment by tmiw** [comment]
> How difficult would be a record from selected microphone to WAV file be to implement @tmiw ?
> 
> I would guess that the nuts and bolts to do this are already present in freedv. Maybe the controls could be in the Options->Audio tab where the voicekeyer setup is already.

Yeah, much of the code is likely already there, just need to figure out how best to show the UI for it.

Though really, perhaps it's a good opportunity to do some cleanup. I'm not sure how frequently the "record from modulator" option's used, for example.

**Comment by DG9BFC** [comment]
sdr console also has a builtin voice keyer (with recording option) .. it records in 16 bit mono 48 kS and you can have up to 50 files maximum (normally not needed but nice to have) ... not only cq calls but also other audio files (that are converted to above numbers) can be played ... be it a test tone signal, a cq call or a jingle or a previously made recording of the other stations signal ... play once or play in a loop (with setable pause in between) ... so most parts are built in freedv .. only needs a record from mic and stop button and save under xyz filename ... voice keyer then should have a drop down list with the filenames
in the setup we have that record button to test the audio input but that does not work to record a longer cq call...
recording from modulator does record the freedv data stream but not the mic audio so we need a button RECORD from mic in (to be used as analog file for cq calls)
dg9bfc sigi
ps in the meantime i use a different way to record a wave file for cq

**Comment by tmiw** [comment]
I'm not sure yet another button on the main window is necessarily a good idea. What if I made the existing Record button record analog audio from the microphone on TX instead of whatever's coming over the air? Any downsides to doing it that way?

**Comment by DG9BFC** [comment]
what about adding another line in the drop down list we have?? (record file from mic in) ... no need to have another button on the direct available frontend (GUI) but we need one to record the mic audio (for recording a cq call)... having that in the drop down list is enough (how often do you need to record a new cq call??) 
-----
but for transmitting that cq call the voice keyer button should open a selection box so we can use "several" recorded wave files to play (open the wav directory, select, press enter)
wdyt??
dg9bfc sigi

**Comment by barjac** [comment]
> ## what about adding another line in the drop down list we have?? (record file from mic in) ... no need to have another button on the direct available frontend (GUI) but we need one to record the mic audio (for recording a cq call)... having that in the drop down list is enough (how often do you need to record a new cq call??)
> 
If the CQ or testing transmission includes the band and other specific mode etc. information then probably more often than you imagine!
It would be nice to be able to record on the fly and use instantly.

> but for transmitting that cq call the voice keyer button should open a selection box so we can use "several" recorded wave files to play (open the wav directory, select, press enter) wdyt?? dg9bfc sigi

No. The selection of the wav in use should not have to be made every time the 'Voice Keyer' button is pressed, it needs to be selectable agreed but not from that button, unless by using a modifier of some sort like SHIFT/click or preferably a right click context menu when hovering over the button. (Mouse only not involving keyboard and two hands). A context menu could also have the record message option. I think it's preferable to have these options available close to the PTT and Voice keyer rather than having to dive off into the main settings menu.

Whilst on the voice keyer topic it seems that the operator has no way to monitor the audio being transmitted other than visually. Could the wav be also piped to the freedv RX audio out channel during voice keyer operation? This would allow the operator to pick a point at which to switch seamlessly to live transmission from within a recording.

**Comment by barjac** [comment]
> I'm not sure yet another button on the main window is necessarily a good idea. What if I made the existing Record button record analog audio from the microphone on TX instead of whatever's coming over the air? Any downsides to doing it that way?

Not really however see my comment above. For new users it may be rather confusing to use the same button for two totally different functions.
Keeping all voice keyer functions grouped in a context menu related to the Voice Keyer button would I think be more logical.

**Comment by Tyrbiter** [comment]
I think that the whole design must not be overcomplicated, so the minimum would be:

Allow 1 or more .wav files to be recorded, perhaps allowing the file name to be edited before saving.

Allow 1 of these .wav files to be played as the Tx audio source, selection of which file could use the suggestion from  @barjac  with a right-click voicekeyer button context menu. This then asks the question about what to do on subsequent voicekeyer use, last file used or file selection dialog? First use after restarting freedv also needs an answer. Save the choice in the .freedv file?

Maybe a visual indication of the audio being sent would be useful, but then it also needs to work with accessibility options.

**Comment by tmiw** [comment]
New proposal:

1. Update main window "Record" button so that it picks up the encoded mic audio when transmitting (as right now it just records RX audio).
    * Additionally, remove the Record from Modulator option from the Tools menu as I don't think that's used much and probably adds confusion.
2. Ctrl/Apple-click on the Voice Keyer button will let you choose an alternate WAV file just for that use of the keyer. (i.e. regular click without keyboard modifiers will continue to have the existing behavior).
3. Add new Tools->Record From Microphone option to allow recording new analog WAV files for use with the voice keyer.
    * Not sure how the main window should behave when this is running. Maybe disable the TX, Record and Voice Keyer buttons but otherwise behave like TX normally does (i.e. switch to the Microphone plot)?

Thoughts?

**Comment by tmiw** [comment]
I just opened PR #493, which contains everything but (3) above (except that right-click triggers the alternate VK file selection instead of Ctrl-click). I'm still not fully sure how to (or whether we even want to) allow recording of new voice keyer files from within the FreeDV application. To be honest, given that FreeDV is a *lot* less picky about WAV files than it used to be, it is conceivable to just have people use Audacity or similar to create VK files and let FreeDV resample as needed. On the other hand, it is pretty expected at this point to be able to record new recordings directly onto radios that have voice keyers, so I can see why people may want something similar in FreeDV.

Anyway, let's make sure the initial set of changes didn't break anything first :)

**Comment by DG9BFC** [comment]
record from mic added in drop down under tools (ok on frontend also nice but having all record options in ONE PLACE is less confusing .. deleting record from modulator sometimes may be usefull so no need to delete it .. ust add record from mic in the menue ald be done with it
safe under (filename)
play that voicekeyer file would be left click (plays with the settings the user wants) .. right click opens a menue where you can select file (and maybe a "pause timer")
hmm??

**Comment by DG9BFC** [comment]
record on frontend should record the decoded audio from demod so that you can sed it to other station by mail or retransit it to him by air

**Comment by DG9BFC** [comment]
typing with fat fingers in a hurry so sorry for typos :-)

**Comment by tmiw** [comment]
For the VK recording, I ended up going with adding a popup menu under the Voice Keyer button when you right-click on the button. There's a new option added in the previously mentioned PR that will put FreeDV into TX and record the microphone audio; pushing PTT again will stop recording.

**Comment by Tyrbiter** [comment]
The Voice Keyer context menu opens off-screen if close to the right edge, is there a way of making it remain inside the freedv main window?

**Comment by DG9BFC** [comment]
what if you use freedv in fullduplex (without ptt button press)??
i use freedv in fulldupex cause i then only need ptt button in the rig (sdr console) but no messing with virtual serial ports, port splitters etc. ... so for me only fullduplex work without lots off switching of serial ports etc
does record also work then?
dg9bfc sigi

**Comment by tmiw** [comment]
> what if you use freedv in fullduplex (without ptt button press)?? i use freedv in fulldupex cause i then only need ptt button in the rig (sdr console) but no messing with virtual serial ports, port splitters etc. ... so for me only fullduplex work without lots off switching of serial ports etc does record also work then? dg9bfc sigi

Yep, I just tested and the VK recorder does indeed work. 👍 I presume the Record button on the left hand side works the same way, too.

**Comment by DG9BFC** [comment]
so .. when you have a release to download .. drop me a note

**Comment by barjac** [comment]
> For the VK recording, I ended up going with adding a popup menu under the Voice Keyer button when you right-click on the button. There's a new option added in the previously mentioned PR that will put FreeDV into TX and record the microphone audio; pushing PTT again will stop recording. Let's see how that goes +1

After stopping recording/transmitting with PTT, does a VK button left click immediately use the new recording? I have not yet tested, but that sounds ideal, as it allows recording an initial CQ and then instant VK repeating of the same initial call.
Nice!

**Comment by tmiw** [comment]
> After stopping recording/transmitting with PTT, does a VK button left click immediately use the new recording? I have not yet tested, but that sounds ideal, as it allows recording an initial CQ and then instant VK repeating of the same initial call. Nice!

It does now. 👍

**Comment by DG9BFC** [comment]
i think recording without going into ptt on state would be better 
you CAN press ptt in parallel but it should be possible to record a cq call without going into ptt on
that would be better i guess
imagine you want to "prepare" a few files .. but want to use them LATER
just thinking ... 
dg9bfc sigi

**Comment by tmiw** [comment]
> i think recording without going into ptt on state would be better you CAN press ptt in parallel but it should be possible to record a cq call without going into ptt on that would be better i guess imagine you want to "prepare" a few files .. but want to use them LATER just thinking ... dg9bfc sigi

The mic is actually only active in TX right now (unless full duplex). I'm not sure how big of a change this would be. Additionally, one advantage of requiring PTT is that it provides a convenient way of showing that the microphone is indeed active (along with a way to stop recording). 

Maybe we can revisit for a future release?

**Comment by DG9BFC** [comment]
as i wrote .. i use freedv in fullduplex with sdr console and 2 virtual cables .. so i anly need to switch the sdr soft into tx mode with ptt button (like in voice) and do not need any vox or messing around with virtual serial ports and port splitting etc ... much easier setup then so i only use fullduplex ... simplex and ptt is nice with a normal rig (ptt in one one serial port .. ptt out on a different port via rts signal .. but for sat work on qo100 fullduplex is much easier

**Comment by tmiw** [comment]
> as i wrote .. i use freedv in fullduplex with sdr console and 2 virtual cables .. so i anly need to switch the sdr soft into tx mode with ptt button (like in voice) and do not need any vox or messing around with virtual serial ports and port splitting etc ... much easier setup then so i only use fullduplex ... simplex and ptt is nice with a normal rig (ptt in one one serial port .. ptt out on a different port via rts signal .. but for sat work on qo100 fullduplex is much easier

Wait, if you're triggering PTT from your SDR program instead of FreeDV itself (and there's no mechanism for FreeDV to know that you kicked off PTT), wouldn't that make the voice keyer kinda painful to use in general?

Anyway, I was doing a bit of research and it looks like the IC-705 (as an example) doesn't put the radio into transmit while you're recording a new voice keyer message. Do you guys have any ideas as to how recording should be indicated/stopped if we make FreeDV not require TX for recording?

**Comment by DG9BFC** [comment]
sure for voice keyer it does not work in fullduplex like i would need it .. there only simplex (and vox) does work
a serial port splitter would solve that but its an additional soft to use and i wanted to keep it simple .. especially cause i try different digi modes
on the sat fullduplex should be used (i guess freedv is no exception)
i also then can check "what the other should hear" ... only thing missing is a mute or attenuate rx while txing so i could have similar level change as i have in analog voice .. when txing the rx level is still there but set a bit lower ... an attenuate rx level (not mute but set down) while txing .. that would be superb (and if we have that for sure i need a serial port routing and normal ptt)

**Comment by tmiw** [comment]
Closing as VK record support has been merged.
